pointerevents/pointerevent_touch-action-keyboard.html

```diff
@@ -78,66 +78,66 @@ <h4 id="desc">Test Description: Press DOWN ARROW key. Wait for description updat
+            async function send_key_to_target(key_code) {
+                return new test_driver.Actions()
+                    .addKeyboard("keyboard1", "keyboard")
+                    .keyDown(key_code, {origin: target0})
+                    .keyUp(key_code, {origin: target0})
+                    .send();
+            }
 
+            async function run() {
                 var test_touchaction = async_test("touch-action attribute test");
-                var actions_promise;
 
-                xScr0 = target0.scrollLeft;
-                yScr0 = target0.scrollTop;
+                const target0 = document.getElementById("target0");
+                const x_scroll_baseline = target0.scrollLeft;
+                const y_scroll_baseline = target0.scrollTop;
 
-                target0.focus();
+                // Inject keyboard scroll inputs.
+                const arrow_down = "\uE015";
+                const arrow_right = "\uE014";
 
                 on_event(target0, 'scroll', function(event) {
-                    xScr1 = target0.scrollLeft;
-                    yScr1 = target0.scrollTop;
-
-                    if(xScr1 != xScr0) {
-                        xScrollIsReceived = true;
-                    }
-
-                    if(yScr1 != yScr0) {
+                    x_scroll = target0.scrollLeft;
+                    y_scroll = target0.scrollTop;
+                    if (y_scroll > y_scroll_baseline && !got_y_scroll) {
                         test_touchaction.step(function () {
-                            yScrollIsReceived = true;
-                            assert_true(true, "y-scroll received.");
+                            assert_true(true, "content was scrolled down.");
                         });
-                        updateDescriptionNextStepKeyboard();
+                        got_y_scroll = true;
                     }
-
-                    if(xScrollIsReceived && yScrollIsReceived) {
-                        // Make sure the test finishes after all the input actions are completed.
-                        actions_promise.then( () => {
-                            test_touchaction.done();
+                    if (x_scroll > x_scroll_baseline && !got_x_scroll) {
+                        test_touchaction.step(function () {
+                            assert_true(true, "content was scrolled right.");
                         });
+                        got_x_scroll = true;
+                    }
+                    if (got_x_scroll && got_y_scroll) {
                         updateDescriptionComplete();
+                        test_touchaction.done();
                     }
                 });
 
-                // Inject keyboard scroll inputs.
-                const arrow_down = "\uE015";
-                const arrow_right = "\uE014";
+                document.addEventListener('keyup', async function(e) {
+                    if (e.code == "ArrowDown") {
```

Comment:
Isn't it simpler to replace L127-130 with a single sequence in L139 as follows?
```
await send_key_to_target(arrow_down);
await send_key_to_target(arrow_right);
await send_key_to_target(arrow_down);
await send_key_to_target(arrow_right);
```

Reply:
Waiting for the arrowdown keyup event is done to deflake the test; without it the test sometimes fails in Safari due to the keypresses firing nearly at the same time. Chrome had flakiness with this test in the past when it used test_driver.Actions() instead of test_driver.sendKeys(), but I have ran the test with my changes hundreds of times in both browsers and have noticed no flakiness in either, which leads me to believe some sort of pause mechanism is necessary for both browsers here.